app/src/org/commcare/android/adapters/EntitySubnodeDetailAdapter.java

```diff
@@ -32,8 +33,24 @@ public EntitySubnodeDetailAdapter(Context context, Detail detail, List<TreeRefer
         this.context = context;
         this.detail = detail;
         this.references = references;
-        this.entities = entities;
-        this.modifier = modifier;
+
+        // Only include data for entities that have at least one valid field
+        this.references = new Vector<>();
+        this.entities = new Vector<>();
+        int entityIndex = 0;
+        for (Entity e : entities) {
+            boolean isValid = false;
+            for (int i = 0; i < e.getNumFields(); i++) {
+                if (e.isValidField(i)) {
+                    isValid = true;
+                }
+            }
+            if (isValid) {
+                this.entities.add(e);
+                this.references.add(references.get(entityIndex));

```

Comment:
The `isValid` boolean isn't needed: if you move these two `add`s into the body of the `e.isValidField(i)` conditional and then break afterwards it should be the same behavior without an intermediate boolean.